metavalidator/any.go

```diff
@@ -0,0 +1,61 @@
+package metavalidator
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+)
+
+var _ tfsdk.AttributeValidator = anyValidator{}
+
+// anyValidator validates that value validates against at least one of the value validators.
+type anyValidator struct {
+	valueValidators []tfsdk.AttributeValidator
+}
+
+// Description describes the validation in plain text formatting.
+func (v anyValidator) Description(ctx context.Context) string {
+	var descriptions []string
+	for _, validator := range v.valueValidators {
+		descriptions = append(descriptions, validator.Description(ctx))
+	}
+
+	return fmt.Sprintf("Value must satisfy at least one of the validations: %s", strings.Join(descriptions, " + "))
+}
+
+// MarkdownDescription describes the validation in Markdown formatting.
+func (v anyValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+// Validate performs the validation.
+// If the diagnostics returned from the validator does not contain an error
+// we return without any diagnostics.
+func (v anyValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
+	for _, validator := range v.valueValidators {
+		validatorResp := &tfsdk.ValidateAttributeResponse{
+			Diagnostics: diag.Diagnostics{},
+		}
+
+		validator.Validate(ctx, req, validatorResp)
+		if !validatorResp.Diagnostics.HasError() {
+			resp.Diagnostics = []diag.Diagnostic{}
```

Comment:
Believe we need https://github.com/hashicorp/terraform-plugin-framework/pull/392 so that we can extract the warnings from diagnostics and return them. Otherwise if we're using `Any()` with `All()` it's possible that we'd have errors from a previous iteration that would be returned, such as the following:
```go
  Any(
    All(validator1, validator2, validator3),
    All(validator4, validator5,validator6),
  )
```

Reply:
My personal semantics opinion here is that given the above A || B, where A may contain conflicting logic than B and where B passes, that returning any warnings from A could potentially be confusing for practitioners. If provider developers want the ability to also return all warnings from other logical branches, then we could either introduce an "options" type parameter to allow triggering the variant behavior, or create a separate variant validator.

If we did want to return warnings from other logical branches, then the logic would need to fully remove any early return behavior and use a variable for tracking whether any given validator was "passing", performing the warnings extraction at the end of function based on that tracking variable to remove error diagnostics to prevent any sort of validator ordering semantics within Any().

Reply:
Have created a separate `AnyWithAllWarningsValidator` but more than happy to change the name  if something else would suit better.